Disallow creating instances of abstract classes

dwblaikie opened the pull request:
A good first-pass, at least. (abstract adapters are rejected with this change, though pending further language design discussion)

dwblaikie committed:
Add more test coverage

dwblaikie committed:
Experimenting with explicit completeness+non-abstractness query

dwblaikie committed:
More experimenting

dwblaikie committed:
More experimenting

dwblaikie committed:
Seems to be mostly working...

Taking notes here:
Newly added diagnostics (+ tested, - untested):
ClassAbstractHere (note, tested in several paths)
AbstractTypeInAdaptDecl+ (not sure if this is correct - is it valid to adapt an abstract type?)
AbstractTypeInConversion- (equivalent IncompleteTypeInConversion is also untested, so I'm not sure if/how to reach this)
AbstractTypeInFunctionParam+
AbstractTypeInFunctionReturnType- (Failures about "initialization of abstract type" seem to block the return type error? That doesn't seem as useful as actually erroring on the return type)
AbstractTypeInInit++
AbstractTypeInLetDecl+
AbstractTypeInValueConversion++
AbstractTypeInVarDecl++ (Field and Variable)

dwblaikie committed:
Simplify and tidy up

dwblaikie committed:
Add TODO

dwblaikie committed:
Cleanup

zygoloid commented on toolchain/check/context.h:
```diff
@@ -333,11 +340,12 @@ class Context {
   // TryToCompleteType. `diagnoser` must not be null.
-  auto AsCompleteType(SemIR::TypeId type_id, BuildDiagnosticFn diagnoser)
+  auto AsCompleteType(SemIR::TypeId type_id, BuildDiagnosticFn diagnoser,
+                      BuildDiagnosticFn abstract_diagnoser)
```
Should we include a default argument for `abstract_diagnoser`? I think it's optional here, right?

dwblaikie replied:
Yep! (allows updating one of the callers that was already passing nullptr)

zygoloid commented on toolchain/check/handle_binding_pattern.cpp:
```diff
@@ -99,16 +99,28 @@ static auto HandleAnyBindingPattern(Context& context, Parse::NodeId node_id,
+            return context.emitter().Build(
+                type_node, AbstractTypeInVarDecl,
+                parent_class_decl ? llvm::StringLiteral("Field")
+                                  : llvm::StringLiteral("Variable"),
```
```suggestion
                parent_class_decl ? llvm::StringLiteral("field")
                                  : llvm::StringLiteral("variable"),
                cast_type_id);
          },
          [&] {
            CARBON_DIAGNOSTIC(AbstractTypeInVarDecl, Error,
                              "{0} has abstract type `{1}`",
                              llvm::StringLiteral, SemIR::TypeId);
            return context.emitter().Build(
                type_node, AbstractTypeInVarDecl,
                parent_class_decl ? llvm::StringLiteral("field")
                                  : llvm::StringLiteral("variable"),
```
Let's fix the case of these while we're here.

zygoloid commented on toolchain/check/handle_class.cpp:
```diff
@@ -393,13 +393,22 @@ auto HandleParseNode(Context& context, Parse::AdaptDeclId node_id) -> bool {
+                          SemIR::TypeId);
+        return context.emitter().Build(node_id, AbstractTypeInAdaptDecl,
+                                       adapted_type_id);
+      });
```
It's not clear to me whether this should be valid. Let's reject for now, but I've filed #4387 to consider this.

zygoloid commented on toolchain/check/testdata/class/fail_abstract.carbon:
```diff
@@ -20,16 +70,59 @@ class Derived {
-  // CHECK:STDERR: fail_abstract.carbon:[[@LINE+3]]:19: error: cannot construct instance of abstract class; consider using `partial Abstract` instead
+  // CHECK:STDERR: fail_abstract.carbon:[[@LINE+7]]:10: error: initialization of abstract type `Abstract`
   // CHECK:STDERR:   return {.base = {.a = 1}, .d = 7};
-  // CHECK:STDERR:                   ^~~~~~~~
```
Is the old diagnostic here still reachable? If not, we should remove it.

dwblaikie replied:
Seems unreachable - removed.

zygoloid commented on toolchain/check/convert.cpp:
```diff
@@ -937,24 +937,46 @@ auto Convert(Context& context, SemIR::LocId loc_id, SemIR::InstId expr_id,
+            CARBON_DIAGNOSTIC(AbstractTypeInInit, Error,
+                              "initialization of abstract type `{0}`",
+                              SemIR::TypeId);
+            CARBON_DIAGNOSTIC(AbstractTypeInValueConversion, Error,
+                              "forming value of abstract type `{0}`",
+                              SemIR::TypeId);
+            CARBON_DIAGNOSTIC(AbstractTypeInConversion, Error,
+                              "invalid use of abstract type `{0}`",
+                              SemIR::TypeId);
```
I think we don't want to diagnose in these two cases, only when initializing.

dwblaikie replied:
Done (this also included adding null `DiagnosticBuilder` - open to riffing on the API (currently just uses the default ctor to create a null DiagnosticBuilder, but could use a named ctor of some kind - and has explicit bool conversion which could be some more explicitly named function))

zygoloid replied:
I think a named constructor would be nice. Maybe even a method on `DiagnosticEmitter`, say `BuildSuppressed` or something, so you don't need to repeat the template argument.

zygoloid commented on toolchain/check/handle_binding_pattern.cpp:
```diff
@@ -185,13 +197,22 @@ static auto HandleAnyBindingPattern(Context& context, Parse::NodeId node_id,
+            return context.emitter().Build(type_node, IncompleteTypeInLetDecl,
+                                           cast_type_id);
+          },
+          [&] {
+            CARBON_DIAGNOSTIC(AbstractTypeInLetDecl, Error,
+                              "`let` binding has abstract type `{0}`",
+                              SemIR::TypeId);
+            return context.emitter().Build(type_node, AbstractTypeInLetDecl,
+                                           cast_type_id);
+          });
```
I think this case should be permitted, as a `let` binding is forming a value not an object.

dwblaikie replied:
Done - thanks for the help understanding the language design here.

zygoloid commented on toolchain/check/handle_function.cpp:
```diff
@@ -364,15 +364,26 @@ static auto HandleFunctionDefinitionAfterSignature(
+                                         IncompleteTypeInFunctionParam,
+                                         param_info.inst.type_id);
+        },
+        [&] {
+          CARBON_DIAGNOSTIC(
+              AbstractTypeInFunctionParam, Error,
+              "parameter has abstract type `{0}` in function definition",
+              SemIR::TypeId);
+          return context.emitter().Build(param_info.inst_id,
+                                         AbstractTypeInFunctionParam,
+                                         param_info.inst.type_id);
+        });
```
I think this case should be permitted because a parameter binding is a value binding (by default, and we don't support the `var` modifier yet).

dwblaikie replied:
Done

zygoloid reviewed (commented):
The intended model is that you cannot create objects of abstract class types, but you can have *values* of abstract class types -- the value representation in that case is a pointer to the instance, so it doesn't need the class to be non-abstract. So I think we would want to remove some of these checks; I'll point out which ones.

dwblaikie committed:
Add default null abstract_diagnoser

dwblaikie committed:
Remove diagnosis for AbstractTypeInConversion and AbstractTypeInValueConversion

This also adds an invalid/null/empty state to DiagnosticBuilder so this
callback can then return the null DiagnosticBuilder for the cases that
don't need errors.

dwblaikie committed:
Add initialization

dwblaikie committed:
Fix case of field/variable in diagnostics

Co-authored-by: Richard Smith <richard@metafoo.co.uk>

dwblaikie committed:
Update tests to match the case of the diagnostic change

dwblaikie committed:
Allow abstract on let declarations.

dwblaikie committed:
Allow abstract params due to them being value bindings

dwblaikie committed:
Remove dead diagnostic for abstract construction

dwblaikie committed:
Coverage for "abstract return type" diagnostic

dwblaikie reviewed (commented):
Handled feedback, and did a more major pass on the test coverage, splitting things (I feel like maybe I /should/ be writing those tests differently to, for example, share the definition of the abstract class rather than redefining it in even split - though it's so short importing it doesn't seem much more legible either - totally open to suggestions there).

Did manage to get test coverage on the return type check (at a call site to, rather than a definition of the abstract-returning function).

zygoloid commented on toolchain/diagnostics/diagnostic_emitter.h:
```diff
@@ -55,6 +55,7 @@ class DiagnosticEmitter {
     DiagnosticBuilder(DiagnosticBuilder&&) noexcept = default;
     auto operator=(DiagnosticBuilder&&) noexcept
         -> DiagnosticBuilder& = default;
+    DiagnosticBuilder() : emitter_(nullptr) {}
```
A brief comment here would be handy.

zygoloid commented on toolchain/check/convert.cpp:
```diff
@@ -963,19 +955,14 @@ auto Convert(Context& context, SemIR::LocId loc_id, SemIR::InstId expr_id,
+            if (!target.is_initializer()) {
+              return DiagnosticEmitter<SemIRLoc>::DiagnosticBuilder();
+            }
+            // return // add support for creating a null DiagnosticBuilder and
+            //  return one here. Include null-testability and use that
+            //  to skip adding notes in the caller if it's null.
```
```suggestion
```
Looks like this has been addressed already.

zygoloid reviewed (approved):
Thanks, looks good.

dwblaikie committed:
Remove temporary comment

Co-authored-by: Richard Smith <richard@metafoo.co.uk>

dwblaikie committed:
Add named ctor and some comments.

dwblaikie committed:
Merge branch 'trunk' into disallow_abstract_instance

dwblaikie committed:
Fix some stuff lost in the merge
